simplify data_parallel.h

jonsneyers opened the pull request:
MSVC was complaining about the function pointers having `const &` qualifiers which are meaningless/implied. I hope this will reduce MSVC warnings (see https://github.com/libjxl/libjxl/issues/933).

Also simplifies things a bit by always returning `bool` (and not sometimes `Status` which gets lost anyway), and not doing the SkipInit() case by means of an explicit specialization with a dummy struct template param.

veluca93 commented:
The change to SkipInit seems good, but why not always return a Status? Why would the Status get lost anyway? (it should have an attribute that makes it an error to do so)

Also, taking InitFunc/DataFunc by value can introduce unnecessary copies which could be a performance footgun.

jonsneyers commented:
> The change to SkipInit seems good, but why not always return a Status? Why would the Status get lost anyway? (it should have an attribute that makes it an error to do so)

The Status gets lost anyway because RunOnPool returns bool.

> Also, taking InitFunc/DataFunc by value can introduce unnecessary copies which could be a performance footgun.

As far as I understand (but my understanding may be wrong), functions are never passed by value, i.e. passing by reference is implicit.

veluca93 commented:
On Tue, 14 Dec 2021, 10:50 Jon Sneyers, ***@***.***> wrote:

> The change to SkipInit seems good, but why not always return a Status? Why
> would the Status get lost anyway? (it should have an attribute that makes
> it an error to do so)
>
> The Status gets lost anyway because RunOnPool returns bool.
>
Then we could make RunOnPool return Status too (perhaps unless it gets
passed skipinit)

> Also, taking InitFunc/DataFunc by value can introduce unnecessary copies
> which could be a performance footgun.
>
> As far as I understand (but my understanding may be wrong), functions are
> never passed by value, i.e. passing by reference is implicit.
>
Yes, but that's not true for lambdas.

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/libjxl/libjxl/pull/1005#issuecomment-993361427>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOPAI4WN3PAF4FZ3MF6FHTUQ4HM7ANCNFSM5KAJXSUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jonsneyers commented:
> Then we could make RunOnPool return Status too (perhaps unless it gets passed skipinit)

Any idea how to do that? I think the easiest way might be to just have a RunOnPoolSkipInit with one argument less (the skipinit is implicit) that returns void, and the one with explicit init returns Status. Does that sound reasonable?

> Also, taking InitFunc/DataFunc by value can introduce unnecessary copies which could be a performance footgun. As far as I understand (but my understanding may be wrong), functions are never passed by value, i.e. passing by reference is implicit.
> Yes, but that's not true for lambdas.

Ah OK. Then I hope the MSVC warnings were just about the `const` and not about the `&`.

veluca93 commented:
On Tue, 14 Dec 2021, 11:00 Jon Sneyers, ***@***.***> wrote:

> Then we could make RunOnPool return Status too (perhaps unless it gets
> passed skipinit)
>
> Any idea how to do that? I think the easiest way might be to just have a
> RunOnPoolSkipInit with one argument less (the skipinit is implicit) that
> returns void, and the one with explicit init returns Status. Does that
> sound reasonable?
>
Yeah that should work. Can also do two overloads, one with skipinit and the
other one for other things, or even fancier things, but those probably need
too much SFINAE for comfort...

> Also, taking InitFunc/DataFunc by value can introduce unnecessary copies
> which could be a performance footgun. As far as I understand (but my
> understanding may be wrong), functions are never passed by value, i.e.
> passing by reference is implicit.
> Yes, but that's not true for lambdas.
>
> Ah OK. Then I hope the MSVC warnings were just about the const and not
> about the &.
>
Just a reference doesn't work either ;)
I suggest we just suppress the warning in msvc.

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/libjxl/libjxl/pull/1005#issuecomment-993370121>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOPAI4YCXIWAWNI32YJPSDUQ4IVLANCNFSM5KAJXSUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jonsneyers commented:
Updated it to keep the `const &`, suppressing the MSVC warnings about it.

Also added a `void RunOnPoolSkipInit` so the one where init isn't skipped can return `Status`. Updated all the places where it is called accordingly.

This will help in case we change some of those init functions to actually fail, e.g. because they couldn't allocate the per-thread storage. There were quite a few places where no-skip-init return values were ignored, and places where skip-init return values were checked (while they're by definition true).

veluca93 commented on lib/jxl/base/data_parallel.h:
```diff
@@ -104,21 +97,34 @@ class ThreadPool {
   if (pool == nullptr) {
     ThreadPool default_pool(nullptr, nullptr);
-    ret = default_pool.Run(begin, end, init_func, data_func, caller);
+    JXL_CHECK(default_pool.Run(begin, end, ThreadPool::TrivialInit, data_func,
```
So... how do we know that the no-init version cannot fail? Does the C API for threading state that this is the case?

jonsneyers replied:
Only the init returns something. The data function has a void return type.

I don't know what should happen if someone implements an incorrect JxlParallelRunner that returns failure even if the init function returned success. This check will then fail.

veluca93 replied:
What if the parallel runner fails for internal reasons?

jonsneyers replied:
It's not clear to me what the actual api is for parallel runners. If it can fail for internal reasons, then we have a bunch of places where this is not handled (none of the calls to RunOnPoolSkipInit check the result). So I was assuming that this cannot happen...

veluca93 replied:
Then the correct solution would be to handle those errors, or to clarify in the documentation of the thread pool API that this should not happen.

jonsneyers replied:
Which option do you prefer? Currently it looks like the thread pool API allows a runner to fail for other reasons than the init failing, but none of our code with SkipInit is handling that possibility.

I'm inclined to say that the thread pool API should be changed so it cannot fail if init succeeds - otherwise it would be a technically valid implementation to just always fail, or to have unreasonable requirements like only working if 256 threads are available. I think it's better to make it so that a thread runner can only return failure if init failed, even if that means it has to fall back to just running things sequentially.

veluca93 commented:
I don't think it would be a big burden to say that a thread pool is allowed
to fail for whatever reason. Of course, if it does fail, then the whole
decoding process will fail too.

On Wed, 15 Dec 2021, 22:46 Jon Sneyers, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In lib/jxl/base/data_parallel.h
> <https://github.com/libjxl/libjxl/pull/1005#discussion_r770063790>:
>
> >    if (pool == nullptr) {
>      ThreadPool default_pool(nullptr, nullptr);
> -    ret = default_pool.Run(begin, end, init_func, data_func, caller);
> +    JXL_CHECK(default_pool.Run(begin, end, ThreadPool::TrivialInit, data_func,
>
> Which option do you prefer? Currently it looks like the thread pool API
> allows a runner to fail for other reasons than the init failing, but none
> of our code with SkipInit is handling that possibility.
>
> I'm inclined to say that the thread pool API should be changed so it
> cannot fail if init succeeds - otherwise it would be a technically valid
> implementation to just always fail, or to have unreasonable requirements
> like only working if 256 threads are available. I think it's better to make
> it so that a thread runner can only return failure if init failed, even if
> that means it has to fall back to just running things sequentially.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/libjxl/libjxl/pull/1005#discussion_r770063790>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOPAI6HVLVOHYR44J3H4YLUREEEFANCNFSM5KAJXSUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jonsneyers commented:
> I don't think it would be a big burden to say that a thread pool is allowed to fail for whatever reason. Of course, if it does fail, then the whole decoding process will fail too.

OK, so I add JXL_RETURN_IF_FAILURE to all the `RunOnPoolSkipInit`s then?

veluca93 commented:
Yeah something like that.

On Wed, 15 Dec 2021, 22:52 Jon Sneyers, ***@***.***> wrote:

> I don't think it would be a big burden to say that a thread pool is
> allowed to fail for whatever reason. Of course, if it does fail, then the
> whole decoding process will fail too.
>
> OK, so I add JXL_RETURN_IF_FAILURE to all the RunOnPoolSkipInits then?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/libjxl/libjxl/pull/1005#issuecomment-995242308>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOPAI7QO3662FSPR7UKQWDUREE2FANCNFSM5KAJXSUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jonsneyers commented:
> Yeah something like that.

OK, did that. This might be a first small step in the direction of being able to detect OOM and gracefully return an error instead of aborting - though we'll need a lot more Status plumbing for that, I guess (and checking if allocation failed).

veluca93 commented on lib/jxl/base/data_parallel.h:
```diff
@@ -104,21 +97,33 @@ class ThreadPool {
+template <class DataFunc>
+Status RunOnPoolSkipInit(ThreadPool* pool, const uint32_t begin,
+                         const uint32_t end, const DataFunc& data_func,
+                         const char* caller) {
```
I believe now we can just have this call RunOnPool without duplicating code :)

(although perhaps at this point I would remove this function entirely...)

jonsneyers replied:
Right. Removed the function.

veluca93 commented on tools/hdr/tone_map.cc:
```diff
@@ -1,89 +1,89 @@
-  image.metadata.m.color_encoding = c_out;
-  JXL_CHECK(jxl::EncodeToFile(image, output_filename, &pool));
-}
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
```
What's up with this file?

jonsneyers replied:
The line endings were CR LF for all files in this directory, so I did a `dos2unix` on all of them because my editor shows the CR visibly as `^M` which annoyed me :)

veluca93 replied:
Yeeeeah... Good idea.
